### PR TITLE
Follow credential generation advice from topology

### PR DESCRIPTION
**Pull request by bbockelm**
This uses the credential generation advice from topology (@jhiemstrawisc - note you may need to do some follow-up to do the same thing for the director code path) when acquiring a token.

The two major changes are determining whether (a) an existing token is compatible with the advice and (b) changes in acquiring the token to generate the correct requested scopes.

Because the token generation logic is in subpackage "oauth2" and the credential generation structure was in the main package, there was a circular dependency.  Hence, the namespace code needed to be split off into a separate sub-package, causing the majority of the code churn you see here.

The following now works:

```
stashcp  /tmp/baz osdf:///chtc/staging/bbockelm/baz.6
```

(for those of you with a CHTC account -- should work similarly for the OSPool issuer)

### CODE REVIEW

**Comment by jhiemstrawisc** [comment]
I tested by downloading a few files from staging and things look good to me. Sometime later this week, I'll start getting things set up director-side to handle the new credential generation info, and then get everything hooked up in the client.

One question I do have is about the naming of max scope depth, both in the client and in topology. To me, the name might imply that a token generated to read from `/chtc/jhiemstra` can read files at most three directories deep, ie `/chtc/jhiemstra/foo/bar/baz`, but no directory below that. At least, that was my initial confusion. Maybe we want something like max scope coarseness?

**Comment by matyasselmeci** [comment]
I think the max scope is the deepest scope that the token will be generated for -- that is, if the max scope is 3 then even if the directory is `/chtc/jhiemstra/foo/bar/baz`, the token will be generated scoped for `/chtc/jhiemstra/foo/` (and everything below).

**Comment by jhiemstrawisc** [comment]
I did figure that out after looking at what the code was actually doing, I'm mostly just opining that the naming confused me until I looked under the hood. What problem is solved by allowing a namespace to specify a max scope depth?

**Comment by matyasselmeci** [comment]
I think that attribute is supposed to be descriptive of what the issuer does rather than what the client should ask the issuer to do; I'd have to ask @bbockelm to learn why we're including it, though.

**Comment by jhiemstrawisc** [comment]
Good to merge?

**Comment by bbockelm** [comment]
@jhiemstrawisc - I think we're ready to merge but you need to approve it as a reviewer first.

The benefit from the max scope depth are:
1. Generating tokens that are more likely to be reused.  This needs guidance from the issuer because it, not `stashcp`, has insight into the layout of the namespace.
2. This is a recommendation from the issuer.  It _could_ be implemented issuer side but there are potential cases (such as generating a shareable URL) where the user agent knows it really wants something more fine-grained.

Regarding the naming -- I think I'm OK with it as-is.  I'm not sure "max scope coarseness" is all that more illuminating than "max scope depth".

File: namespaces/namespaces.go
Comment (jhiemstrawisc) [inline namespaces/namespaces.go:1]: With the new CredentialGeneration struct, we wind up with the issuer being stored twice (both in the new struct, and in its encapsulating Namespace struct). Do we need both, or can we remove one of them?

**Review by jhiemstrawisc — APPROVED** [review]
LGTM.

**Pull request merged**